### PR TITLE
feat: added queue manage logic for records

### PR DESCRIPTION
**Pull request by CL-SHLOMIKONCHA**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+*Import & Initialize*
+
+
+
+```
+from mc-pyority-queue.task_handler import TaskHandler
+
+task_handler = TaskHandler('job_type', 'task_type',
+    'http://localhost:8081', 'http://localhost:8080/heartbeat', 1.0)
```
Comment (asafMasa) [inline README.md:36]: why 1.0? do you support 0.5?

File: README.md [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+```
+above example uses TaskHanler class and initializing it follow the request params:
+
+`task_handler = TaskHandler(job_type, task_type, host_base_url, heartbeat_url, interval_ms)`
```
Comment (asafMasa) [inline README.md]: host_base_url ->job_mngr_url

File: README.md [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+await task_hanlder.reject()
+```
+ack handle completed task - stops sending task's heartbeat and handle complete task status
+
+<br />
+
+* **update_progress**
+
+```
+await task_hanlder.update_progress()
```
Comment (asafMasa) [inline README.md]: params?

File: src/heartbeat.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+        self.thread = None
+        self.interval_ms = interval_ms
+
+    async def start(self, task_id):
+        try:
+            heartbeat_url = f'{self.base_url}/{task_id}'
+            self.thread = threading.Thread(target=self.thread_callback, args=[heartbeat_url])
+            self.thread.start()
+        except Exception as e:
+            print(f'Error occurred: {e}.')  # TODO: replace by mc-logger
```
Comment (asafMasa) [inline src/heartbeat.py]: add logger for all 'print' methods

File: src/heartbeat.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+            raise e
+
+    def stop(self):
+        try:
+            if self.running is True:
+                self.running = False
+                print('stopping thread')  # TODO: replace by mc-logger
```
Comment (asafMasa) [inline src/heartbeat.py]: better log on what thread is being closed (Stopping heartbeat)

File: src/heartbeat.py [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+                print('thread was stopped ')  # TODO: replace by mc-logger
+        except Exception as e:
+            print(f'Error occurred: {e}.')  # TODO: replace by mc-logger
+            raise e
+
+    async def send_heartbeat(self, url, interval_ms):
+        try:
+            while self.running is True:
+                await asyncio.sleep(interval_ms)
+                print(f'sending heartbeat to {url}')  # TODO: replace by mc-logger
```
Comment (asafMasa) [inline src/heartbeat.py]: don't need a log

File: src/heartbeat.py [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+                async with aiohttp.ClientSession() as session:
+                    async with session.post(url) as response:
+                        print("Status:", response.status)  # TODO: replace by mc-logger
```
Comment (asafMasa) [inline src/heartbeat.py]: no need for log here

File: src/heartbeat.py [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+                        print("Content-type:", response.headers['content-type'])  # TODO: replace by mc-logger
+                        await response.json()
+        except Exception as e:
+            print(f'Error occurred: {e}.')  # TODO: replace by mc-logger
+            raise e
```
Comment (asafMasa) [inline src/heartbeat.py]: don't stop the task on heartbeat failure

File: src/records.py [hunk 9]
```diff
@@ -0,0 +1,57 @@
+import aiohttp
+from heartbeat import *
+
+
+class Records:
+    def __init__(self, job_type, task_type):
+        self.job_type = job_type
+        self.task_type = task_type
+
+    async def get(self, base_url, job_id):
```
Comment (asafMasa) [inline src/records.py]: base_url -> move to contructor

File: README.md [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+ await task_hanlder.reject()
+```
+reject handle error by stop sending task's heartbeat and handle task's status - depends if task is recoverable or not.
+
+<br />
+
+* **ack**
+
+```
+await task_hanlder.reject()
```
Comment (asafMasa) [inline README.md]: ack

File: README.md [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+```
+
+consume task from the job manager service and start send heartbeat to the heartbeat service
+
+<br />
+  
+* **Reject**
+
+```
+ await task_hanlder.reject()
```
Comment (asafMasa) [inline README.md]: add an argument of recovarable (bool)

File: src/records.py [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+        try:
+            get_task_url = f'{base_url}/job/{job_id}/tasks'
+            async with aiohttp.ClientSession() as session:
+                print(f'GET Request')  # TODO: replace by mc-logger
+                async with session.get(get_task_url) as response:
+                    print("Status:", response.status)  # TODO: replace by mc-logger
```
Comment (asafMasa) [inline src/records.py]: if error messages are handled in the except section then remove this message

File: src/records.py [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+            async with aiohttp.ClientSession() as session:
+                print(f'POST Request')  # TODO: replace by mc-logger
+                headers = {'accept': 'application/json', 'Content-Type': 'application/json'}
+                async with session.post(dequeue_url, headers=headers) as response:
+                    print("Status:", response.status)  # TODO: replace by mc-logger
+                    print("Content-type:", response.headers['content-type'])  # TODO: replace by mc-logger
+                    resp = await response.json()
+                    return resp
+        except Exception as e:
+            print(f'Error occurred: {e}.')  # TODO: replace by mc-logger
```
Comment (asafMasa) [inline src/records.py]: don't print 404 as it is a valid scenario (empty queue)

File: src/records.py [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+            raise e
+
+    async def update(self, base_url, job_id, task_id, payload):
+        try:
+            update_url = f'{base_url}/job/{job_id}/tasks/{task_id}'
+            async with aiohttp.ClientSession() as session:
+                print(f'PUT Request')  # TODO: replace by mc-logger
```
Comment (asafMasa) [inline src/records.py]: better log

File: src/records.py [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+                headers = {'accept': 'application/json', 'Content-Type': 'application/json'}
+                async with session.put(update_url, json=payload, headers=headers) as response:
+                    print("Status:", response.status)  # TODO: replace by mc-logger
```
Comment (asafMasa) [inline src/records.py]: if error messages are handled in the except section then remove this message

File: src/task_handler.py [hunk 16]
```diff
@@ -0,0 +1,61 @@
+from heartbeat import *
+from records import *
+from enums.statuses import Statuses
+
+
+class TaskHandler:
+    def __init__(self, job_type, task_type, host_base_url, heartbeat_url, interval_ms):
```
Comment (asafMasa) [inline src/task_handler.py]: rename host_base_url to job_manager_url

File: src/records.py (hunk 9 above)
Comment (asafMasa) [inline src/records.py]: get is not being used what is it's purpose?

File: src/records.py [hunk 17]
```diff
@@ -0,0 +1,57 @@
+import aiohttp
+from heartbeat import *
+
+
+class Records:
+    def __init__(self, job_type, task_type):
```
Comment (asafMasa) [inline src/records.py]: add base_url to contructor

File: src/task_handler.py [hunk 18]
```diff
@@ -0,0 +1,61 @@
...
+        self.heartbeat = Heartbeat(self.heartbeat_url, interval_ms)
+
+    async def dequeue(self):
+        try:
+            resp = await self.record.consume(self.host_base_url, self.job_type, self.task_type)
+            task_id = resp['id']
+            await self.heartbeat.start(task_id)
+        except Exception as e:
+            print(f'Error occurred: {e}.')  # TODO: replace by mc-logger
+            raise e
```
Comment (asafMasa) [inline src/task_handler.py:24]: separate the 404 from other exceptions

File: src/task_handler.py [hunk 19]
```diff
@@ -0,0 +1,61 @@
...
+
+    async def reject(self, job_id, task_id, recoverable):
+        try:
+            self.heartbeat.stop()
+            if recoverable is True:
+                payload = {
+                    'status': Statuses.PENDING.value,
+                    'attempts': 1  # TODO: replace with increment value ***
```
Comment (asafMasa) [inline src/task_handler.py]: need to fix this

**Review by asafMasa — CHANGES_REQUESTED** [review]
Open issue on retries
Good job

File: README.md (hunk 1 above)
Comment (CL-SHLOMIKONCHA) [reply README.md:36]: yes, it is support 0.5 or 0.001, 01, etc..

File: README.md (hunk 3 above)
Comment (CL-SHLOMIKONCHA) [reply README.md]: fixed in 
37b4bbe

File: README.md (hunk 2 above)
Comment (CL-SHLOMIKONCHA) [reply README.md]: fixed in 
37b4bbe

File: src/task_handler.py (hunk 19 above)
Comment (CL-SHLOMIKONCHA) [reply src/task_handler.py]: fixed in 559c61a

File: src/task_handler.py (hunk 18 above)
Comment (CL-SHLOMIKONCHA) [reply src/task_handler.py:24]: fixed in 559c61a

File: src/records.py (hunk 17 above)
Comment (CL-SHLOMIKONCHA) [reply src/records.py]: fixed in 559c61a

File: src/records.py (hunk 9 above)
Comment (CL-SHLOMIKONCHA) [reply src/records.py]: fixed in 
559c61a

File: src/task_handler.py (hunk 16 above)
Comment (CL-SHLOMIKONCHA) [reply src/task_handler.py]: fixed in 
559c61a

File: src/heartbeat.py (hunk 5 above)
Comment (CL-SHLOMIKONCHA) [reply src/heartbeat.py]: fixed in 
4c17adc

File: src/heartbeat.py (hunk 4 above)
Comment (CL-SHLOMIKONCHA) [reply src/heartbeat.py]: fixed in 
77f8c3a

File: src/logger/json_logger.py [hunk 20]
```diff
@@ -0,0 +1,22 @@
+from MapColoniesJSONLogger.logger import generate_logger
```
Comment (asafMasa) [inline src/logger/json_logger.py]: get logger from outside

**Review by asafMasa — CHANGES_REQUESTED** [review]

**Review by asafMasa — APPROVED** [review]

File: src/logger/json_logger.py (hunk 20 above)
Comment (CL-SHLOMIKONCHA) [reply src/logger/json_logger.py]: fixed in c7807cb

**Pull request merged**